userspace/tudumvc/updating-the-agent.md

```diff
@@ -0,0 +1,544 @@
+
+Add some todos and mark at least one of them as complete, then open the browser's console (`F12` in most browsers).
+
+In the console, let's examine localStorage, as that's what TodoMVC is currently using to store data:
+```
+>> localStorage
+<- >Storage { todos: "[{\"done\":false,\"id\":\"d160cc1a-02dc-4ea3-f89c-43b482da5fc4\",\"label\":\"two\"},{\"done\":false,\"id\":\"f555c8b9-ca31-b25e-d71e-f473c5de38f6\",\"label\":\"one\"},{\"done\":false,\"id\":\"a8297d73-3359-7829-cdc4-c545642f9a35\",\"label\":\"test\"}]", length: 1 }
+```
+`localStorage` is storing an object with a key of "todos" whose value is an array of data from our TodoMVC todo list. Take a closer look using:
+```
+>> JSON.parse(localStorage.getItem("todos"))
+<- (3) […]
+0: Object { done: false, id: "d160cc1a-02dc-4ea3-f89c-43b482da5fc4", label: "two" }
+1: Object { done: false, id: "f555c8b9-ca31-b25e-d71e-f473c5de38f6", label: "one" }
+2: Object { done: true, id: "a8297d73-3359-7829-cdc4-c545642f9a35", label: "test" }
+length: 3
+```
+Using the `parse` method of JSON to turn the "todos" item from `localStorage` into its JSON form, you should an array of objects each with three key-value pairs: (1) "done" - the completion status, (2) "id" - a random ID used to differentiate tasks, (3) "label" - the task text. This structure should inform our Urbit's required state upgrades. `%tudumvc` will need to acccommodate storing tasks in a way that includes the same information points provided by the app natively.
+
+On the Urbit side, you'll further structure the original design of the Earth app's data, turning this array into a `map`. A `map` will enable to indexing a given task quickly by it's "id" to make changes to it (mark it as complete, edit the "label", etc.).
+
+### Upgrading the state {#lesson-state-upgrades}
+First, you'll need to open the `/sur/tudumvc.hoon` file and define our state as a type so that we can reference it easily in our /app file (not necessary, but idiomatic - one could define the state structures in /app, as well). Recall that we import `/sur/tudumvc.hoon` using `/-` in our /app file.
+
+#### `/sur/tudumvc.hoon`
+Our current file only defines our `action` type. For `%tudumvc` to communicate well with TodoMVC (Earth app), you'll want to add a new type called `tasks` that is structured to accommodate a `map` with a key of a task's "id" and a value for that key of the task's "label" and "done"-ness (completion state). Using this structure will allow us to index a given task by its "id" (a unique key) to modify the "label" or "done"ness (which can be non-unique - we can enter "Call Mom" 100 times and not have an indexing conflict - btw, you should call your mother - it's almost Mother's Day).
+
+Just as with `action`, we're going to use [`+$`](https://urbit.org/docs/reference/hoon-expressions/rune/lus/#lusbuc):
+<table>
+<tr>
+<td>
+:: initial version
```

Comment:
This is a comment for most of these two versions of code side by side thing. It's a good idea but with our current site layout it makes both sections nearly unreadable. Maybe it should be something like the tabbed option here? https://urbit.org/getting-started/